File: string/reverse-vowels-of-a-string.py
# https://leetcode.com/problems/reverse-vowels-of-a-string/
# tags - string
class Solution:
    def reverseVowels(self, s: str) -> str:
        chars = list(s)
        vowels = set('aeiouAEIOU')
        vowel_indices = [i for i, c in enumerate(chars) if c in vowels]
        for i in range(len(vowel_indices) // 2):
            temp = chars[vowel_indices[i]]
            chars[vowel_indices[i]] = chars[vowel_indices[-i-1]]
            chars[vowel_indices[-i-1]] = temp
        return ''.join(chars)
        

File: string/reverse-vowels-of-a-string.js
// https://leetcode.com/problems/reverse-vowels-of-a-string/
// tags - string
/**
 * @param {string} s
 * @return {string}
 */
var reverseVowels = function (s) {
  const vowels = new Set("aeiouAEIOU");
  const chars = Array.from(s);
  const vchars = chars.filter((c) => vowels.has(c));
  for (let i = 0, j = vchars.length - 1; i < chars.length; i++) {
    if (vowels.has(chars[i])) {
      chars[i] = vchars[j];
      j--;
    }
  }
  return chars.join("");
};
